### PR TITLE
[1pt] PR: SRC calibration with ras2fim xsection rating data

### PR DESCRIPTION
**Pull request by RyanSpies-NOAA**
This feature branch includes new functionality to perform an additional layer of HAND SRC calibration using ras2fim rating curve and point data. The calibration workflow for ras2fim data follows the same general logic as the existing USGS rating curve calibration routine. Address #907 

### Additions  

- `src/src_adjust_ras2fim_rating.py`: New python script to perform the data prep steps for running the SRC calibration routine: 
1) merge the `ras_elev_table.csv` data and the ras2fim cross section rating curve data (`reformat_ras_rating_curve_table.csv`) 
2) sample the ras2fim rating curve at NWM recurrence flow intervals (2, 5, 10, 25, 50, 100yr)
3) pass inputs to the `src_roughness_optimization.py` workflow

### Changes

- `config/deny_branches.lst`: Added `ras_elev_table.csv` to keep list. Needed for `fim_post_processing.sh`
- `config/deny_unit.lst`: Added `ras_elev_table.csv` to keep list. Needed for `fim_post_processing.sh`
- `config/params_template.env`: Added new block for ras2fim SRC calibration parameters (can turn on/off each of the three SRC calibration routines individually); also reconfigured docstrings for calibration parameters)
- `fim_post_processing.sh`: Added routines to create ras2fim calibration data and then run the SRC calibration workflow with ras2fim data
- `src/add_crosswalk.py`: Added placeholder variable (`calb_coef_ras2fim`) in all `hydrotable.csv` files
- `src/aggregate_by_huc.py`: Added new blocks to perform huc-branch aggregation for all `ras_elev_table.csv` files
- `src/run_by_branch.sh`: Revised input variable (changed from csv file to directory) for `usgs_gage_crosswalk.py` to facilitate both `usgs_elev_table.csv` and ras_elev_table.csv` outputs
- `src/run_unit_wb.sh`: Revised inputs and output variables for `usgs_gage_unit_setup.py` and `usgs_gage_crosswalk.py`
- `src/src_roughness_optimization.py`: Added code blocks to ingest ras2fim rating curve data; added new attributes/renamed output variables to catchments gpkg output
- `src/usgs_gage_crosswalk.py`: Added code block to process ras2fim point locations alongside existing USGS gage point locations; outputs a separate csv if ras2fim points exist within the huc
- `src/usgs_gage_unit_setup.py`: Added code block to ingest and process raw ras2fim point locations gpkg file (same general workflow to usgs gages); all valid points (USGS and RAS2FIM) are exported to the huc level `usgs_subset_gages.gpkg`
- `tools/inundate_nation.py`: Added functionality to allow user to pass in a single HUC for faster spot checking of NWM recurr inundation maps

### Testing

- Feature branch completed successfully (no errors) using the UAT huc list.
- Changes from this feature branch are currently only applicable (ras2fim rating data availability) for 12030105, 12030202, 12040101, 12090301

### Notes
Overview of the 'fim_post_processing.sh' workflow with the addition of `src_adjust_ras2fim_rating.py`
FIM4 Post Processing Calibration Diagram

Summary of the SRC calibration methodology using the ras2fim rating curve data
image

New variables added to all `hydrotable.csv` files: `calb_coef_ras2fim`
image

Map showing 12090301 `gw_catchments_reaches_filtered_addedAttributes_crosswalked_0.gpkg` with symbology representing the `final_calb_coef`
image

Catchment map for 12090301 showing which calibration data source was used in the `final_calb_coef`. Note the use of both spatial data (AHPS libraries) and ras2fim
image

## Results
Rating curve comparison graphics showing the ras2fim cross section rating, HAND pre-calb rating, HAND post-calb rating, and estimated BLE flood extent elevation
image
image

Alpha evaluation metrics comparing 4.0.12.0 vs. the HAND SRC calibration with the ras2fim rating data
image
image

Agreement raster example showing an example of the improved agreement from the HAND SRC with ras2fim rating calibration
image

### Issuer Checklist (For developer use)

_You may update this checklist before and/or after creating the PR. If you're unsure about any of them, please ask, we're here to help! These items are what we are going to look for before merging your code._

- [x] Informative and human-readable title, using the format: `[_pt] PR: <description>`
- [x] Links are provided if this PR resolves an issue, or depends on another other PR
- [x] If submitting a PR to the `dev` branch (the default branch), you have a descriptive [Feature Branch](https://www.atlassian.com/git/tutorials/comparing-workflows/feature-branch-workflow) name using the format: `dev-<description-of-change>` (e.g. `dev-revise-levee-masking`)
- [x] Changes are limited to a single goal (no scope creep)
- [x] The feature branch you're submitting as a PR is up to date (merged) with the latest `dev` branch
- [ ] Changes adhere to [PEP-8 Style Guidelines](https://peps.python.org/pep-0008/)
- [x] Any _change_ in functionality is tested
- [ ] Passes all unit tests locally (inside interactive Docker container, at `/foss_fim/`, run: `pytest unit_tests/`)
- [x] New functions are documented (with a description, list of inputs, and expected output)
- [x] Placeholder code is flagged / future todos are captured in comments
- [x] [CHANGELOG](/docs/CHANGELOG.md) updated with template version number, e.g. `4.x.x.x` 
- [x] [Reviewers requested](https://help.github.com/articles/requesting-a-pull-request-review/)
- [x] Add yourself as an [assignee](https://docs.github.com/en/issues/tracking-your-work-with-issues/assigning-issues-and-pull-requests-to-other-github-users) in the PR  as well as the FIM Technical Lead

### Merge Checklist (For Technical Lead use only)

- [ ] Update [CHANGELOG](/docs/CHANGELOG.md) with latest version number
- [ ] Update the [Citation.cff](/CITATION.cff) file to reflect the latest version number in the [CHANGELOG](/docs/CHANGELOG.md)
- [ ] If applicable, update [README](/README.md) with major alterations

### CODE REVIEW

File: config/params_template.env [hunk 1]
```diff
@@ -44,12 +44,19 @@ export src_subdiv_toggle="True" # Toggle to run composite roughness src routine
 export vrough_suffix="" # text to append to output log and hydrotable file names (use for testing/debugging)
 export vmann_input_file="/data/inputs/rating_curve/variable_roughness/mannings_global_06_12.csv" # input file location with nwm feature_id and channel roughness and overbank roughness attributes
 
-#### apply SRC adjustments using USGS rating curve database (calibration db) ####
-export src_adjust_usgs="True" # Toggle to run src adjustment routine (True=on; False=off)
```
Comment (robgpita-noaa) [inline config/params_template.env:48]: `src_adjust_usgs` is still referenced on line 131 in `fim_post_processing.sh`. If we delete it here, then the call to `src_adjust_usgs_rating.py` on line 136 will not take place. Not sure if this is intentional or not.

Comment (rspies) [reply config/params_template.env:48]: `src_adjust_usgs` is defined below (line 51).  I did some reorganization of these calibration parameters, but I don't think I deleted anything.

Comment (robgpita-noaa) [reply config/params_template.env:48]: Good deal, when previously viewing this PR, my screen didn't show line 51. Thanks!

**Comment by CarsonPruitt-NOAA** [comment]
Really interesting stuff going on, but here's the stacked bar alpha results. I only ran 12090301, but I think it's worth expanding to the other HUCs. The version `fim_4_3_11_0` has spatial calibration and the others either have no calibration or only USGS/ras2fim rating curve calibration.
image
image

To me, the results seem to be very mixed with regard to the NWS LiDAR FIM comparison, possibly with an upper end bias. The BLE comparison jives with your slides above.

**Review by CarsonPruitt-NOAA — APPROVED** [review]

**Comment by CarsonPruitt-NOAA** [comment]
Here's the NWS results from all four ras2fim HUCs.
image

**Pull request merged**